### DISCUSSION TITLE
Creation/Usage of Autogenerated Variables

### DISCUSSION THREAD

**Discussion by MikeLeimon** [idaholab/moose#19538] [Q&A General] [answered]
Greetings All,

I am working on a project where I have a set of known species which I must track (for now let's say it's on the order of ten different species). If possible, I would like to auto-generate variables for all of these species and have an expectation that they exist within every block. Recently, I explored using Actions to auto-generate these variables for all of these species and I think I was able to do this successfully. 

The first issue I have is that **I am not sure how to connect kernels to these auto-generated variables**. Furthermore, I have situations in which one species type can transform into one of the other species types. **How would I go about accessing those auto-generated variables from within an interface?**

At the end of the day, this effort (should I be successful) will result in a drastic simplification to the model inputs. Fundamentally, I suspect the MOOSE framework is likely capable of accomplishing the workflow I am intending, however I am not experienced enough to know how to approach this. 

I do apologize if my questions here have been overly vague, but I very much appreciate any assistance you can provide.

Many thanks,
~Mike

**Comment by loganharbour** [comment, marked as answer]
An easy example to look at in regards to adding both variables and kernels from an action is in https://github.com/idaholab/moose/blob/next/test/src/actions/AddLotsOfDiffusion.C.

This is a test object that simply adds a ton of variables and for each variable adds a diffusion kernel and a boundary condition. I think this should be enough to send you in the right direction, and I'm glad to answer more questions once you get a better idea.

Can you comment more on what you mean about species transformation?

**Reply by MikeLeimon** [reply]
Thank you so much for the quick response!

As an example of a species transformation. Lets say that some species I was tracking were: H, H2, H2O, O, and O2. A species transformation (as I described above) could be characterized as: 2H + O -> H2O. This would result in a decrease in the species variable values for H and O, and an increase in the species variable value for H2O.

**Reply by loganharbour** [reply]
Ahh, duh. I was thinking much more general. Sorry about that.

I think what I'm looking for more then is in regards to this comment: 

> How would I go about accessing those auto-generated variables from within an interface?

What kind of interface, what does it do? What are you trying to do to the variables? etc

**Reply by MikeLeimon** [reply]
Right now I'm focusing on a Sievert interface. Ultimately, I have a case where I have a vapor phase species which can move/transform into a different species within an adjacent block. This is of course the first step though. When it is all said and done, there will be lots of different transformations between the entire set of autogenerated species. Given all the book-keeping that will need to be done, I view this part as being both very tedious and error prone if not done in an automated fashion. 

For now, I'm going to spend a bit of time looking at the example you mentioned earlier, I'll be sure to drop more questions for you should they come up. Also, thanks again for your feedback. It is very much appreciated.

**Reply by loganharbour** [reply]
Absolutely, glad to help. 

Admittedly, I'm not very familiar of what we do in the framework for these physics but have you taken a look at the chemical reactions physics module? https://mooseframework.inl.gov/modules/chemical_reactions/

I'm not sure if you're just trying to solve ODEs or more, but said module should at least be able to give you suggestions on how to manage species in a generalized manner.
